vespalib/src/vespa/vespalib/util/shared_string_repo.h

```diff
@@ -0,0 +1,193 @@
+            uint32_t operator()(const AltKey &key) const { return key.hash; }
+        };
+        struct Equal {
+            const std::vector<Entry> &entries;
+            Equal(const std::vector<Entry> &entries_in) : entries(entries_in) {}
+            Equal(const Equal &rhs) = default;
+            bool operator()(const Key &a, const Key &b) const { return (a.idx == b.idx); }
+            bool operator()(const Key &a, const AltKey &b) const { return ((a.hash == b.hash) && (entries[a.idx].str == b.str)); }
+        };
+        using HashType = vespalib::hash_set<Key,Hash,Equal>;
+
+    private:
+        SpinLock              _lock;
+        std::vector<Entry>    _entries;
+        std::vector<uint32_t> _free;
+        HashType              _hash;
+
+        uint32_t make_entry(const AltKey &alt_key) {
+            if (_free.empty()) {
+                uint32_t idx = _entries.size();
+                _entries.emplace_back(alt_key);
+                return idx;
+            } else {
+                uint32_t idx = _free.back();
+                _free.pop_back();
+                _entries[idx].reuse(alt_key);
+                return idx;
+            }
+        }
+
+    public:
+        Partition()
+            : _lock(), _entries(), _free(), _hash(0, Hash(), Equal(_entries)) {}
+        ~Partition();
+
+        uint32_t resolve(const AltKey &alt_key) {
+            std::lock_guard guard(_lock);
+            auto pos = _hash.find(alt_key);
+            if (pos != _hash.end()) {
+                ++_entries[pos->idx].ref_cnt;
+                return pos->idx;
+            } else {
+                uint32_t idx = make_entry(alt_key);
+                _hash.insert(Key{idx, alt_key.hash});
+                return idx;
+            }
+        }
+
+        vespalib::string get(uint32_t idx) {
+            std::lock_guard guard(_lock);
+            return _entries[idx].str;
+        }
+
+        void copy(uint32_t idx) {
+            std::lock_guard guard(_lock);
+            ++_entries[idx].ref_cnt;
+        }
+
+        void reclaim(uint32_t idx) {
+            std::lock_guard guard(_lock);
+            Entry &entry = _entries[idx];
+            if (--entry.ref_cnt == 0) {
+                _hash.erase(Key{idx, entry.hash});
+                entry.reset();
+                _free.push_back(idx);
+            }
+        }
+    };
+
+    static std::array<Partition,NUM_PARTS> _partitions;
+
+    static uint32_t resolve(vespalib::stringref str) {
+        if (!str.empty()) {
+            uint64_t full_hash = XXH3_64bits(str.data(), str.size());
+            uint32_t part = full_hash & PART_MASK;
+            uint32_t local_hash = full_hash >> PART_BITS;
+            uint32_t local_idx = _partitions[part].resolve(AltKey{str, local_hash});
+            return (((local_idx << PART_BITS) | part) + 1);
+        } else {
+            return 0;
+        }
+    }
+
+    static vespalib::string get(uint32_t id) {
+        uint32_t part = (id - 1) & PART_MASK;
```

Comment:
don't you need to check for `id == 0` here?

Reply:
yes, I do